### ISSUE TITLE
Add astype method to Tensor

### ISSUE THREAD

**Issue by rsokl** [rsokl/MyGrad#115]
How don't we have this?? [Numpy ref](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.astype.html)

**Comment by rsokl** [comment]
Not sure exactly how to handle this. I'm thinking that it should produce a Tensor without a creator - i.e. the resulting tensor wont backprop to anything. 

The alternative is that the present tensor is reproduced with the desired data-type, and retains its creator as well.

Thoughts @davidmascharka ?

**Comment by davidmascharka** [comment]
I think the logic we want is to have it be a copy so that it wouldn't backprop anything. Otherwise you run into the issue of back-converting the gradient. From PyTorch's docs for `to()`:

> If the `self` Tensor already has the correct` torch.dtype` and `torch.device`, then `self` is returned. Otherwise, the returned tensor is a copy of `self` with the desired `torch.dtype` and `torch.device`.

which I think we could do but always return a copy to be consistent

**Comment by rsokl** [comment]
> but always return a copy to be consistent

Yep, I agree. People should never see `.astype` as something that they can backprop through. We should leverage the [`valid_dtype` machinery](https://github.com/rsokl/MyGrad/blob/ba51c03ba0a2e7209c49f0f1343f2e077e20e92b/src/mygrad/tensor_base.py#L162-L166) here ... and that should be updated to exclude complex numbers :sweat_smile:
